api/query_test.go

```diff
@@ -405,6 +421,127 @@ func TestFenceErrorHTTPResponse(t *testing.T) {
 	}
 }
 
+func TestMetadataHappyHTTPResponse(t *testing.T) {
+	testcases := []metadataTest{
+		{
+			name:   "Valid GET Query",
+			method: http.MethodGet,
+			metadata: testMetadataQuery{
+				Vds: well_known,
+				Sas: "n/a",
+			},
+			expectedStatus: http.StatusOK,
+		},
+		{
+			name:   "Valid json POST Query",
+			method: http.MethodPost,
+			metadata: testMetadataQuery{
+				Vds: well_known,
+				Sas: "n/a",
+			},
+			expectedStatus: http.StatusOK,
+		},
+	}
+
+	for _, testcase := range testcases {
+		w := httptest.NewRecorder()
+		ctx, r := gin.CreateTestContext(w)
+		setupTestServer(r)
+
+		prepareMetadataRequest(ctx, t, testcase)
+		r.ServeHTTP(w, ctx.Request)
+
+		if w.Result().StatusCode != testcase.expectedStatus {
+			msg := "Got status %v; want %d %s in case '%s'"
+			t.Errorf(
+				msg,
+				w.Result().Status,
+				testcase.expectedStatus,
+				http.StatusText(testcase.expectedStatus),
+				testcase.name,
+			)
+		}
+		metadata := w.Body.String()
+		expectedMetadata := `{
+			"axis": [
+				{"annotation": "Inline", "max": 5.0, "min": 1.0, "samples" : 3, "unit": "unitless"},
+				{"annotation": "Crossline", "max": 11.0, "min": 10.0, "samples" : 2, "unit": "unitless"},
+				{"annotation": "Sample", "max": 16.0, "min": 4.0, "samples" : 4, "unit": "ms"}
+			],
+			"format": 3
+		}`
+
+		if metadata != expectedMetadata {
+			msg := "Metadata not equal in case '%s'"
+			require.JSONEq(t, expectedMetadata, metadata, fmt.Sprintf(msg, testcase.name))
+		}
+	}
+}
+
+func TestMetadataErrorHTTPResponse(t *testing.T) {
+	testcases := []metadataTest{
+		{
+			name:           "Invalid json GET query",
+			method:         http.MethodGet,
+			metadataQuery:  "help I am a duck",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "invalid character",
+		},
+		{
+			name:           "Invalid json POST query",
+			method:         http.MethodPost,
+			metadataQuery:  "help I am a duck",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "invalid character",
+		},
+		{
+			name:           "Missing parameters GET query",
+			method:         http.MethodGet,
+			metadataQuery:  "{\"vds\":\"" + well_known + "\"}",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Error:Field validation for 'Sas'",
+		},
+		{
+			name:           "Missing parameters POST Query",
+			method:         http.MethodPost,
+			metadataQuery:  "{\"sas\":\"somevalidsas\"}",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Error:Field validation for 'Vds'",
+		},
+		{
+			name:   "Query which passed all input checks but still should fail",
+			method: http.MethodPost,
+			metadata: testMetadataQuery{
+				Vds: "unknown",
+				Sas: "n/a",
+			},
```

Comment:
Why do you use `metadataQuery` in some cases and `metadata` in others?

Reply:
Because metadata is nice and simple struct, visible values, and metadata query is ugly inline json.

I want to test ugly bad inline json in each call because in the code itself we have duplicated code for that (and I don't think it can be unified). For this I can't use a struct, because unmarshalling won't create invalid json.

For missing parameters tests I now can actually use the struct because we uncoupled the test structs and code structs. Tests structs no longer have "required" binding, so it is possible now! Didn't notice before. Would look nicer to me, not sure about you.

Reply:
> I want to test ugly bad inline json in each call because in the code itself we have duplicated code for that (and I don't think it can be unified). For this I can't use a struct, because unmarshalling won't create invalid json.

I think that the class of "invalid json" test cases should be it's own test: TestInvalidJsonBody 

>For missing parameters tests I now can actually use the struct because we uncoupled the test structs and code structs.

Just to clarify, if you construct a struct without a particular field, that field will be default constructed to its type's default value. Marshalling/Unmarshalling works fine, regardless of the `required` tag. Which means you can construct the test case and fire of the request just fine. The struct validator in our request handlers checks that the unmarshalled struct have no field with default values.

TL;DR you can even use the "code" struct in the test case when testing missing fields. This will work:

```
		{
			name:   "Query which passed all input checks but still should fail",
			method: http.MethodPost,
			fence: testFenceQuery{
				Vds:              "unknown",
				CoordinateSystem: "ilxl",
				Coordinates:      [][]float32{{3, 12}, {2, 10}},
				Sas:              "n/a",
			},
			expectedStatus: http.StatusInternalServerError,
			expectedError:  "Could not open VDS",
		},
```

>  Would look nicer to me, not sure about you.

I would definitely prefer to keep json-strings to a minimum wherever possible